### PR TITLE
fix: ensure storages are force flushed during uplink shutdown

### PR DESCRIPTION
**Pull request by de-sh**
Closes #

### Why?

Storage might contain publishes that aren't enough to overflow from in-memory buffer onto disk, in this situation, during shutdown, uplink must force them to be written into file to save them from being lost.

### Trials Performed

Consult linked PRs for more detail

### CODE REVIEW

**Comment by de-sh** [comment]
Let's consider this situation with regards to full-graceful shutdown:
> 1. Uplink is running on device, shutdown triggered, inflight file created with 10 unacked(`MAX_INFLIGHT_SIZE`) and 10 in_channel(Capacity of channel between `AsyncClient` and `EventLoop`) publishes persisted into it.
> 2. Uplink restarts, loads 10 publishes into channel(because of cap and considering `EventLoop` ain't handling requests), rest 10 publishes still left unhandled as network is down.
> 3. Now if uplink were to shutdown again before getting network, would we want to lose the 10 packets that were not pushed onto channel or duplicate the 10 packets that were pushed into channel?
This is something that has been bugging me for a while.

Illustrating the scenario, consider 10 unacked publishes are (A, B) and 10 in_channel publishes are (C, D), first shutdown of uplink writes them to inflight file as (A, B, C, D), on restart only first 10(i.e. A, B) publishes are loaded while rest 10(C, D) are not. The two possible ways to handle next shutdown would be:
1. lose last 10 and be left with only first 10(i.e. on disk: A, B)
2. Duplicate first 10(i.e. on disk: A, B, C, D, A, B)

NOTE: those 10 publishes in_channel after restart will now avalanche if we had uplink restarts continuously(A,B,C,D,A,B,A,B...), while only 10 packets will remain lost in the other scenario(A,B). The second path is what I have taken in the current implementation, but wanted to document this decision and ask for feedback.

**Comment by de-sh** [comment]
Code to ease readability of control handles as requested by @tekjar is included in the PR https://github.com/bytebeamio/uplink/pull/316 and [this commit](https://github.com/bytebeamio/uplink/compare/shutdown-handles...more-handles?expand=1)

File: uplink/src/base/serializer/mod.rs [hunk 1]
```diff
@@ -296,7 +350,6 @@ impl<C: MqttClient> Serializer<C> {
         }
 
         loop {
```
Comment (de-sh) [inline uplink/src/base/serializer/mod.rs:352]:
```suggestion
        loop {
            // Collect next data packet and write to disk
```

**Review by de-sh — COMMENTED** [review]
Remove unnecessary changes

File: uplink/src/base/serializer/mod.rs [hunk 2]
```diff
@@ -278,9 +302,39 @@ impl<C: MqttClient> Serializer<C> {
...
         })
     }
 
+    pub fn ctrl_tx(&self) -> CtrlTx {
+        CtrlTx { inner: self.ctrl_tx.clone() }
+    }
+
+    /// Write all data received, from here-on, to disk only, shutdown serializer
+    /// after handling all data payloads.
+    async fn shutdown(&mut self) -> Result<Status, Error> {
```
Comment (tekjar) [inline uplink/src/base/serializer/mod.rs:316]: Remove async

File: uplink/src/main.rs [hunk 3]
```diff
@@ -200,7 +200,7 @@ fn main() -> Result<(), Error> {
         uplink.resolve_on_shutdown().await.unwrap();
         info!("Uplink shutting down...");
         // NOTE: wait 5s to allow serializer to write to network/disk
-        sleep(Duration::from_secs(5)).await;
+        sleep(Duration::from_secs(10)).await;
```
Comment (tekjar) [inline uplink/src/main.rs:203]: Make shutdown reactive/synchronous in a separate PR

**Pull request merged**